File: 0018-4sum/0018-4sum.py
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        if len(nums)<4:
            return []
        nums.sort()
        vec = []
        for i in range(len(nums)-3):
            if i>0 and nums[i] == nums[i-1]:
                continue
            
            for j in range(i+1, len(nums)-2):
                if j>i+1 and nums[j] == nums[j-1]:
                    continue
                k=j+1
                l=len(nums)-1
                new_target = target - nums[i]-nums[j]
                while(k<l):
                    
                    if nums[k] + nums[l] < new_target:
                        k += 1
                    elif nums[k] + nums[l] > new_target:
                        l -= 1
                    else:
                        vec.append([nums[i], nums[j], nums[k], nums[l]])
                        while k<l and nums[k] == nums[k+1]:
                            k += 1
                        while k<l and nums[l] == nums[l-1]:
                            l -= 1

                        k += 1
                        l -= 1

            
        
        return vec

        

File: 0018-4sum/0018-4sum.cpp
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        if(nums.size()<4){
            return {};
        }

        sort(nums.begin(), nums.end());

        vector<vector<int>> vec;

        for(int i=0; i< nums.size()-3; i++){
            if(i>0 && nums[i]==nums[i-1]){
                continue;
            }
            
            for(int j=i+1; j<nums.size()-2; j++){
                if(j>i+1 && nums[j]==nums[j-1]){
                    continue;
                }
                int k=j+1;
                int l = nums.size()-1;
                long long latest_target =  (long long)target - nums[i] - nums[j];

                while(k<l){
                    if(nums[k]+nums[l] > latest_target){
                        l--;
                    }
                    else if(nums[k]+nums[l] < latest_target){
                        k++;
                    }
                    else{
                        vec.push_back({nums[i], nums[j], nums[k], nums[l]});
                        while(k<l && nums[k]==nums[k+1]) k++;
                           
                        
                        while(k<l && nums[l]==nums[l-1]) l--;
                            
                        k++;
                        l--;
                    }
                }
            }
        }
        return vec;
        
    }
};